### ISSUE TITLE
Que does not seem to pick up jobs, doesn't want to terminate

### ISSUE THREAD

**Issue by jvdp** [chanks/que#213]
I'm using Rails 5 with ActiveJob and am trying to see if I can replace Sidekiq with Que. 

`config.active_job.queue_adapter = :que` in `application.rb`.

When I run `bundle exec que --log-level debug` and queue a job (from the Rails console) I get no output, but development log gets a line added:
```
{"lib":"que","hostname":"xPrime","pid":2486,"thread":70188353976520,"event":"locker_start","queues":["default"]}
```

(Nothing about the job I just queue'd.)

When I try to shutdown the worker with ctrl+c it logs:
```
{"lib":"que","hostname":"xPrime","pid":2486,"thread":70188353976520,"event":"locker_stop"}
```

But the worker doesn't stop; output: `Finishing Que's current jobs before exiting...`.

Any ideas where to start?

**Comment by chanks** [comment]
Hmm, are you logging from inside the job? Would be helpful to see whether it's being picked up.

You can also run the `que` executable with the `--log-internals` option, that should output much more information about what's going on in Que.

**Comment by jvdp** [comment]
When I add logging inside the job it doesn't show up anywhere. With `--log-internals` I get a lot of polling spam with the following in between:
```
I, [2018-03-28T16:14:18.051067 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"locker_polling","object_id":70306908214020,"t":"2018-03-28T14:14:18.051002Z","priorities":{"10":9},"held_locks":[],"queue":"default"}
I, [2018-03-28T16:14:18.051193 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_waiting","object_id":70306903712380,"t":"2018-03-28T14:14:18.051151Z","backend_pid":5685,"channel":"que_listener_5685","timeout":0.05}
I, [2018-03-28T16:14:18.091476 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_received_notification","object_id":70306903712380,"t":"2018-03-28T14:14:18.091345Z","channel":"que_listener_5685","backend_pid":5685,"source_pid":5708,"payload":"{\"message_type\":\"job_available\",\"queue\":\"default\",\"priority\":50,\"id\":2,\"run_at\":\"2018-03-28T14:14:18.080469Z\"}"}
I, [2018-03-28T16:14:18.091679 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_received_messages","object_id":70306903712380,"t":"2018-03-28T14:14:18.091632Z","backend_pid":5685,"channel":"que_listener_5685","messages":[{"message_type":"job_available","queue":"default","priority":50,"id":2,"run_at":"2018-03-28T14:14:18.080469Z"}]}
I, [2018-03-28T16:14:18.092011 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_filtered_messages","object_id":70306903712380,"t":"2018-03-28T14:14:18.091939Z","backend_pid":5685,"channel":"que_listener_5685","messages":[{"message_type":"job_available","queue":"default","priority":50,"id":2,"run_at":"2018-03-28T14:14:18.080469Z"}]}
I, [2018-03-28T16:14:18.092256 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"locker_locking","object_id":70306908214020,"t":"2018-03-28T14:14:18.092210Z","backend_pid":5685,"ids":[2]}
I, [2018-03-28T16:14:18.093122 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"connection_execute","object_id":70306903715520,"t":"2018-03-28T14:14:18.092947Z","backend_pid":5685,"command":"            WITH jobs AS (SELECT * FROM que_jobs WHERE id IN (2))\n            SELECT * FROM jobs WHERE pg_try_advisory_lock(id)\n","params":[],"ntuples":1}
I, [2018-03-28T16:14:18.094105 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"connection_execute","object_id":70306903715520,"t":"2018-03-28T14:14:18.094044Z","backend_pid":5685,"command":"            SELECT id\n            FROM public.que_jobs\n            WHERE finished_at IS NULL\n              AND expired_at IS NULL\n              AND id IN (2)\n","params":[],"ntuples":1}
I, [2018-03-28T16:14:18.094352 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"job_buffer_push","object_id":70306908213000,"t":"2018-03-28T14:14:18.094213Z","maximum_size":8,"ids":[2],"current_queue":[]}
I, [2018-03-28T16:14:18.094730 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"locker_polling","object_id":70306908214020,"t":"2018-03-28T14:14:18.094643Z","priorities":{"10":8},"held_locks":[2],"queue":"default"}
I, [2018-03-28T16:14:18.094834 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_waiting","object_id":70306903712380,"t":"2018-03-28T14:14:18.094800Z","backend_pid":5685,"channel":"que_listener_5685","timeout":0.05}
I, [2018-03-28T16:14:18.146936 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"locker_polling","object_id":70306908214020,"t":"2018-03-28T14:14:18.146847Z","priorities":{"10":8},"held_locks":[2],"queue":"default"}
I, [2018-03-28T16:14:18.147103 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_waiting","object_id":70306903712380,"t":"2018-03-28T14:14:18.147043Z","backend_pid":5685,"channel":"que_listener_5685","timeout":0.05}
I, [2018-03-28T16:14:18.197668 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"locker_polling","object_id":70306908214020,"t":"2018-03-28T14:14:18.197583Z","priorities":{"10":8},"held_locks":[2],"queue":"default"}
I, [2018-03-28T16:14:18.197894 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_waiting","object_id":70306903712380,"t":"2018-03-28T14:14:18.197848Z","backend_pid":5685,"channel":"que_listener_5685","timeout":0.05}
I, [2018-03-28T16:14:18.250308 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"locker_polling","object_id":70306908214020,"t":"2018-03-28T14:14:18.250127Z","priorities":{"10":8},"held_locks":[2],"queue":"default"}
I, [2018-03-28T16:14:18.253055 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"poller_polled","object_id":70306903694720,"t":"2018-03-28T14:14:18.252960Z","queue":"default","locked":0,"priorities":{"10":8},"held_locks":[2],"newly_locked":[]}
I, [2018-03-28T16:14:18.253240 #5625]  INFO -- : {"lib":"que","hostname":"xPrime","pid":5625,"thread":70306908199860,"internal_event":"listener_waiting","object_id":70306903712380,"t":"2018-03-28T14:14:18.253163Z","backend_pid":5685,"channel":"que_listener_5685","timeout":0.05}
```

So it seems to lock but then... nothing?

**Comment by chanks** [comment]
It's weird that it's only looking for jobs of priority 10 or under. Did you set the `--worker-priorities` option when you ran `que`? If so, the job you enqueued wouldn't be run by that worker since it has a priority of 50 (priority is unix-style, so 1 is high priority, 50 is lower priority, etc). Ideally the job with priority 50 wouldn't be distributed to that worker in the first place - I suppose that's an oversight on my part, I was kind of assuming that all workers would have the same priority setup.

**Comment by texpert** [comment]
Please, help!! I am experiencing this trouble too - for 2 days already. Jobs are enqueued, but not executed. Que version was 0.14.1, has upgraded to 0.14.3. Rails was 5.1.5, just upgraded it to 5.1.6

```
[ec2-user@ip-10-0-41-162 current]$ bundle exec rails c
Loading staging environment (Rails 5.1.6)
irb(main):003:0* Que.wake!
=> nil
irb(main):006:0* Que.job_stats
=> [{"queue"=>"", "job_class"=>"ActiveJob::QueueAdapters::QueAdapter::JobWrapper", "count"=>491, "count_working"=>0, "count_errored"=>0, "highest_error_count"=>0, "oldest_run_at"=>2018-03-28 07:00:00 +0000}]
irb(main):007:0> Que.wake!
=> nil
irb(main):008:0> Que.mode = :off
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":3502340,"event":"mode_change","value":"off"}
=> :off
irb(main):009:0> Que.mode = :sync
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":3502340,"event":"mode_change","value":"sync"}
=> :sync
irb(main):010:0> Que.wake!
=> nil
irb(main):011:0> Que.mode = :async
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":3502340,"event":"mode_change","value":"async"}
=> :async
irb(main):012:0> {"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":72887460,"event":"job_unavailable"}
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":72887660,"event":"job_unavailable"}

irb(main):015:0* {"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":72887660,"event":"job_unavailable"}
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":72887660,"event":"job_unavailable"}
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":72887660,"event":"job_unavailable"}
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":72887660,"event":"job_unavailable"}
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":72887660,"event":"job_unavailable"}

irb(main):016:0* Que.mode = :sync
{"lib":"que","hostname":"ip-10-0-41-162","pid":4642,"thread":3502340,"event":"mode_change","value":"sync"}
=> :sync
irb(main):019:0* Que.wake!
=> #<Que::Worker:0x0000000008b05ba0 @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x0000000008b05b00>, @queue="", @state=:working, @thread=#<Thread:0x0000000008b05ad8@/var/app/current/vendor/bundle/ruby/2.3/gems/que-0.14.3/lib/que/worker.rb:19 run>>
```

**Comment by texpert** [comment]
Didn't help. Seems it's time to migrate to 1.0-beta

**Comment by jvdp** [comment]
> It's weird that it's only looking for jobs of priority 10 or under. Did you set the --worker-priorities option when you ran que? If so, the job you enqueued wouldn't be run by that worker since it has a priority of 50 (priority is unix-style, so 1 is high priority, 50 is lower priority, etc). Ideally the job with priority 50 wouldn't be distributed to that worker in the first place - I suppose that's an oversight on my part, I was kind of assuming that all workers would have the same priority setup.

No, I didn't change it but I did run it with just 1 worker. So that would get priority 10 and not pick up anything over? But it's strange that it hangs when I tell it to quit.

**Comment by jvdp** [comment]
Running it with a worker priority of 100 (`bin/que --log-level debug --log-internals -w 1 --worker-priorities 100`) doesn't seem to change anything. The job I am adding is priority 50.

**Comment by chanks** [comment]
Well, I see why running it with just one worker would break things, because that single worker would only work jobs of priority 10 or lower (with the default priorities), so that'd explain the polling query.

What if you run it with the default settings (without specifying a worker count or priorites), does it work then? Or are you running just one worker because you're concerned about the thread-safety of your code?

I think it'll take a bit of work to get 1.0 working well in the single-worker case, because aside from the default priority issue meaning that that worker has a max priority of 10, it also means that the internal job buffer size (which defaults to a max of 8) would likely be way too large, and would mean that some jobs just sit around in some workers' buffers while other workers are idle. Maybe the buffer size default should scale with the number of workers...

**Comment by NARKOZ** [comment]
@chanks I have exactly the same issue while running it with worker count 6. Que version is 1.0.0.beta from master (30081cb), Rails 5.1.5.

**Comment by chanks** [comment]
Hmm, alright, well if there's a problem with Rails in general it could be a lot of things :disappointed: I'll try to find some time to replicate/debug it.

**Comment by nnc** [comment]
@chanks what would be any potential downsides to having `que` establish its own connection based off of ActiveRecord connection config?

**Comment by chanks** [comment]
One hasn't occurred to me yet...

**Comment by chanks** [comment]
Hey y'all, I just made this tweak (so the listener now always spins up its own dedicated connection). I couldn't see any downsides, and it makes the code a bit cleaner anyway, so why not.

I confirmed that this fixed my local reproduction of the error, so I just released 1.0.0.beta2 with this fix. Please let me know if there are any more problems!

**Comment by jvdp** [comment]
@chanks thanks! I will check it out.
